synapse/config/_util.py

```diff
@@ -61,3 +63,20 @@ def json_error_to_config_error(
         else:
             path.append(str(p))
     return ConfigError(e.message, path)
+
+
+class DomainGlobSet:
+    def __init__(self, globs: list):
+        self._globs = []
+        for entry in globs:
+            try:
+                self._globs.append(glob_to_regex(entry.encode("ascii").decode("ascii")))
+            except UnicodeEncodeError:
+                raise ConfigError("IDNA domain names are not allowed: %s " % (entry,))
+
+    @lru_cache(maxsize=100000)
```

Comment:
suggest adding a comment here recording your notes about how `@lrucache` is good enough for now, but a future optimisation could be to use a tree.